File: results/0.0/SQLITE3-DB/sqlite3-db-gpt-4o.c
#include <stdio.h>
#include <stdlib.h>
#include <hammer/hammer.h>

HParser *create_sqlite3_parser() {
    // Define parsers for each field in the SQLite3 header
    HParser *header_string = h_token("SQLite format 3\000", 16);
    HParser *page_size = h_uint16();
    HParser *write_version = h_uint8();
    HParser *read_version = h_uint8();
    HParser *reserved_space = h_uint8();
    HParser *max_embedded_payload_fraction = h_uint8();
    HParser *min_embedded_payload_fraction = h_uint8();
    HParser *leaf_payload_fraction = h_uint8();
    HParser *file_change_counter = h_uint32();
    HParser *database_size_in_pages = h_uint32();
    HParser *first_freelist_trunk_page = h_uint32();
    HParser *total_freelist_pages = h_uint32();
    HParser *schema_cookie = h_uint32();
    HParser *schema_format_number = h_uint32();
    HParser *default_page_cache_size = h_uint32();
    HParser *largest_root_btree_page_number = h_uint32();
    HParser *text_encoding = h_uint32();
    HParser *user_version = h_uint32();
    HParser *incremental_vacuum_mode = h_uint32();
    HParser *application_id = h_uint32();
    HParser *reserved_for_expansion = h_repeat_n(h_uint8(), 20);
    HParser *version_valid_for_number = h_uint32();
    HParser *sqlite_version_number = h_uint32();

    // Combine all parsers into a single parser for the SQLite3 header
    HParser *sqlite3_header = h_sequence(
        header_string,
        page_size,
        write_version,
        read_version,
        reserved_space,
        max_embedded_payload_fraction,
        min_embedded_payload_fraction,
        leaf_payload_fraction,
        file_change_counter,
        database_size_in_pages,
        first_freelist_trunk_page,
        total_freelist_pages,
        schema_cookie,
        schema_format_number,
        default_page_cache_size,
        largest_root_btree_page_number,
        text_encoding,
        user_version,
        incremental_vacuum_mode,
        application_id,
        reserved_for_expansion,
        version_valid_for_number,
        sqlite_version_number,
        NULL
    );

    return sqlite3_header;
}

void parse_sqlite3_file(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Failed to open file");
        exit(EXIT_FAILURE);
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    unsigned char *buffer = malloc(file_size);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(file);
        exit(EXIT_FAILURE);
    }

    fread(buffer, 1, file_size, file);
    fclose(file);

    HParser *parser = create_sqlite3_parser();
    HParseResult *result = h_parse(parser, buffer, file_size);

    if (result) {
        printf("SQLite3 file parsed successfully.\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse SQLite3 file.\n");
    }

    free(buffer);
    h_parser_free(parser);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <sqlite3-db-file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    parse_sqlite3_file(argv[1]);
    return EXIT_SUCCESS;
}

File: results/0.0/SQLITE3-DB/sqlite3-db-gpt-4o.rs
use nom::{
    bytes::complete::{tag, take},
    number::complete::{be_u16, be_u32, be_u64, be_u8},
    IResult,
};
use std::fs::File;
use std::io::Read;
use std::env;

#[derive(Debug)]
struct SqliteHeader {
    header_string: [u8; 16],
    page_size: u16,
    file_format_write_version: u8,
    file_format_read_version: u8,
    reserved_space: u8,
    max_embedded_payload_fraction: u8,
    min_embedded_payload_fraction: u8,
    leaf_payload_fraction: u8,
    file_change_counter: u32,
    database_size: u32,
    first_freelist_trunk_page: u32,
    total_freelist_pages: u32,
    schema_cookie: u32,
    schema_format_number: u32,
    default_page_cache_size: u32,
    largest_root_btree_page: u32,
    text_encoding: u32,
    user_version: u32,
    incremental_vacuum_mode: u32,
    application_id: u32,
    reserved: [u8; 20],
    version_valid_for: u32,
    sqlite_version_number: u32,
}

fn parse_sqlite_header(input: &[u8]) -> IResult<&[u8], SqliteHeader> {
    let (input, header_string) = take(16usize)(input)?;
    let (input, page_size) = be_u16(input)?;
    let (input, file_format_write_version) = be_u8(input)?;
    let (input, file_format_read_version) = be_u8(input)?;
    let (input, reserved_space) = be_u8(input)?;
    let (input, max_embedded_payload_fraction) = be_u8(input)?;
    let (input, min_embedded_payload_fraction) = be_u8(input)?;
    let (input, leaf_payload_fraction) = be_u8(input)?;
    let (input, file_change_counter) = be_u32(input)?;
    let (input, database_size) = be_u32(input)?;
    let (input, first_freelist_trunk_page) = be_u32(input)?;
    let (input, total_freelist_pages) = be_u32(input)?;
    let (input, schema_cookie) = be_u32(input)?;
    let (input, schema_format_number) = be_u32(input)?;
    let (input, default_page_cache_size) = be_u32(input)?;
    let (input, largest_root_btree_page) = be_u32(input)?;
    let (input, text_encoding) = be_u32(input)?;
    let (input, user_version) = be_u32(input)?;
    let (input, incremental_vacuum_mode) = be_u32(input)?;
    let (input, application_id) = be_u32(input)?;
    let (input, reserved) = take(20usize)(input)?;
    let (input, version_valid_for) = be_u32(input)?;
    let (input, sqlite_version_number) = be_u32(input)?;

    Ok((
        input,
        SqliteHeader {
            header_string: header_string.try_into().unwrap(),
            page_size,
            file_format_write_version,
            file_format_read_version,
            reserved_space,
            max_embedded_payload_fraction,
            min_embedded_payload_fraction,
            leaf_payload_fraction,
            file_change_counter,
            database_size,
            first_freelist_trunk_page,
            total_freelist_pages,
            schema_cookie,
            schema_format_number,
            default_page_cache_size,
            largest_root_btree_page,
            text_encoding,
            user_version,
            incremental_vacuum_mode,
            application_id,
            reserved: reserved.try_into().unwrap(),
            version_valid_for,
            sqlite_version_number,
        },
    ))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <sqlite3_db_file>", args[0]);
        return;
    }

    let filename = &args[1];
    let mut file = File::open(filename).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_sqlite_header(&buffer) {
        Ok((_, header)) => {
            println!("{:?}", header);
        }
        Err(e) => {
            eprintln!("Failed to parse SQLite header: {:?}", e);
        }
    }
}